### ISSUE TITLE
Performance Hit in Background

### ISSUE THREAD

**Issue by juliand665** [LambdAurora/LambdaBetterGrass#11]
Hi! I'm the developer of [Dynamic FPS](https://github.com/juliand665/Dynamic-FPS), a mod which throttles Minecraft's rendering when unfocused or hidden to improve performance of the rest of the system. I just noticed that I wasn't seeing nearly as much benefit in with background anymore, and out of all the mods I'm using (including Sodium, Indium, CBT, …), LambdaBetterGrass turned out to be the culprit, raising background CPU usage from <5% to ~70%.

~~Dynamic FPS has a pretty simple approach to throttling rendering: it [hooks into `GameRenderer`](https://github.com/juliand665/Dynamic-FPS/blob/main/src/main/java/dynamicfps/mixin/GameRendererMixin.java) and cancels `render` if not enough time (1 second by default) has elapsed since the last render (`checkForRender()` also [force-idles](https://github.com/juliand665/Dynamic-FPS/blob/c727b69f2abdeb82145d69fe7a1b8c994a2545fa/src/main/java/dynamicfps/DynamicFPSMod.java#L96) for a short time to avoid a busy loop). Do you have any ideas what might be causing so much computation when paused and not rendering? It seems to happen even in the main menu, even when hidden (i.e. never rendering at all).~~

It turns out this happens with or without Dynamic FPS. Without, background CPU usage goes from 25-30% to 85-90% (out of one core) on my machine with macOS 11. Since LambdAurora can't reproduce on her end, I still have yet to investigate whether it's machine-bound or OS-bound.

**Comment by juliand665** [comment]
I was using 1.0.2 (in 1.16) and I just noticed you've since release 1.0.3; I'll test with that and update this.

**Comment by juliand665** [comment]
Confirmed to still happen with 1.0.3, in 1.16.5

**Comment by LambdAurora** [comment]
That is quite weird.

In LBG there's nothing running in the background, nor in SpruceUI.

Rendering stuff should only happen when MC asks for it, so I'm not sure what is happening here.

**Comment by juliand665** [comment]
Yeah, it's very strange. We don't really mix into the same places either, right? Perhaps we can narrow this down somewhat by testing with only SpruceUI, or disabling parts of your mod or something? I don't really know how it's structured so this is just a thought, but perhaps just disabling some of your mixins would work to narrow it down?

**Comment by LambdAurora** [comment]
The 2 mixins are only called at resource loading time, one to inject the replaced models (and read the metadata files of LBG), and the other one to inject the generated textures.

SpruceUI should only intervene in the config GUI, but we can still test.

**Comment by PanoptesDreams** [comment]
> Hi! I'm the developer of [Dynamic FPS](https://github.com/juliand665/Dynamic-FPS)...
...and out of all the mods I'm using (including Sodium, Indium, CBT,"

> ...mods I'm using (including... *CBT*..."

*cough* excuse me WHAT mod? Our research only led to depravity ☠
Google isn't what it used to be... 😓

**Comment by juliand665** [comment]
Connected Block Textures, iirc

**Comment by PanoptesDreams** [comment]
> Connected Block Textures, iirc

For future reader reference here's how you get C*** B*** T****** in Minecraft under Fabric, [Continuity](https://www.curseforge.com/minecraft/mc-mods/continuity)
